op_crates/http/lib.rs

```diff
@@ -0,0 +1,408 @@
+  rid: ResourceId,
+  _data: Option<ZeroCopyBuf>,
+) -> Result<ResourceId, AnyError> {
+  let http_server = state
+    .borrow()
+    .resource_table
+    .get::<HttpServer>(rid)
+    .ok_or_else(bad_resource_id)?;
+
+  let (tcp_stream, _) = http_server.listener.accept().await?;
+  let deno_service = DenoService::default();
+
+  let hyper_connection = Http::new()
+    .with_executor(LocalExecutor)
+    .serve_connection(tcp_stream, deno_service.clone());
+
+  let conn_resource = ConnResource {
+    hyper_connection: Rc::new(RefCell::new(hyper_connection)),
+    deno_service,
+  };
+  let rid = state.borrow_mut().resource_table.add(conn_resource);
+
+  Ok(rid)
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NextRequestResponse(
+  // connection_closed:
+  bool,
+  // request_body_rid:
+  Option<ResourceId>,
+  // response_sender_rid:
+  ResourceId,
+  // method:
+  String,
+  // headers:
+  Vec<(String, String)>,
+  // url:
+  String,
+);
+
+pub async fn op_next_request(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  _data: Option<ZeroCopyBuf>,
+) -> Result<NextRequestResponse, AnyError> {
+  let conn_resource = state
+    .borrow()
+    .resource_table
+    .get::<ConnResource>(rid)
+    .ok_or_else(bad_resource_id)?;
+
+  poll_fn(|cx| {
+    let poll_result =
+      conn_resource.hyper_connection.borrow_mut().poll_unpin(cx);
+
+    if let Poll::Ready(Err(e)) = poll_result {
+      return Poll::Ready(Err(e));
+    }
+
+    if let Some(request_resource) =
+      conn_resource.deno_service.inner.borrow_mut().take()
+    {
+      let tx = request_resource.response_tx;
+      let req = request_resource.request;
+      let method = req.method().to_string();
+
+      let mut headers = Vec::with_capacity(req.headers().len());
+      for (name, value) in req.headers().iter() {
+        let name = name.to_string();
+        let value = value.to_str().unwrap_or("").to_string();
+        headers.push((name, value));
+      }
+
+      let url = req.uri().to_string();
+
+      let has_body = if let Some(exact_size) = req.size_hint().exact() {
+        exact_size > 0
+      } else {
+        true
+      };
+
+      let maybe_request_body_rid = if has_body {
+        let stream: BytesStream = Box::pin(req.into_body().map(|r| {
+          r.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+        }));
+        let stream_reader = StreamReader::new(stream);
+        let mut state = state.borrow_mut();
+        let request_body_rid = state.resource_table.add(RequestBodyResource {
+          reader: AsyncRefCell::new(stream_reader),
+          cancel: CancelHandle::default(),
+        });
+        Some(request_body_rid)
+      } else {
+        None
+      };
+
+      let mut state = state.borrow_mut();
+      let response_sender_rid =
+        state.resource_table.add(ResponseSenderResource(tx));
+
+      let req_json = NextRequestResponse(
+        matches!(poll_result, Poll::Ready(Ok(()))),
```

Comment:
This is a bit tricky and I'm pretty sure we've handled this situation wrong. If polling the connection returns `Poll::Ready(Ok(())` it means that the connection has been closed. If so it doesn't really make sense to return request to JS because we won't be able to respond to it anyway (because the connection is already closed). @ry thoughts?

Reply:
Not all requests are necessarily about the response. For some requests you care more about the request than the response. I think we should pass all requests we have received into the isolate, even if the user won't be able to respond to them.

Reply:
Yeah, a good example use-case is "fire and forget" logging, where users just care about pushing an event into a queue